utxo/src/cache.rs

```diff
@@ -36,9 +36,53 @@ pub struct ConsumedUtxoCache {
     pub(crate) best_block: Id<GenBlock>,
 }
 
-#[derive(Clone)]
+// We don't use std::borrow::Cow because that requires ToOwned to be implemented,
+// and we don't use MaybeOwned crate because that doesn't offer the ?Sized trait,
+// which means that trait objects won't work with it
+pub enum UtxosViewCow<'a> {
+    Borrowed(&'a dyn UtxosView),
+    Owned(Box<dyn UtxosView + 'a>),
+}
+
+impl<'a> UtxosView for UtxosViewCow<'a> {
+    fn utxo(&self, outpoint: &OutPoint) -> Option<Utxo> {
+        match self {
+            UtxosViewCow::Borrowed(r) => r.utxo(outpoint),
+            UtxosViewCow::Owned(r) => r.utxo(outpoint),
```

Comment:
What do you think about implementing something like that in order to avoid matching in every method?
```rust
impl<'a> UtxosViewCow<'a> {
    fn as_ref(&self) -> &dyn UtxosView {
        match self {
            UtxosViewCow::Borrowed(r) => *r,
            UtxosViewCow::Owned(o) => o.as_ref(),
        }
    }
}
```